### PR TITLE
Add Llama 3 tutorial (porting HF model to JAX)

### PR DESCRIPTION
**Pull request by windmaple**

### CODE REVIEW

File: docs/JAX_porting_HF_model.md [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+
+# Converting the LLama 3 8B Instruct model from Hugging Face to JAX
+
+[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/jax-ml/jax-ai-stack/blob/main/docs/convert_from_HF.ipynb)
+
+[![Kaggle](https://kaggle.com/static/images/open-in-kaggle.svg)](https://github.com/jax-ml/jax-ai-stack/blob/main/docs/JAX_for_LLM_pretraining.ipynb)
+
+This tutorial demonstrates how to you could convert Meta's [Llama 3 8B Instruct model](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct) from Hugging Face to a JAX model and run it on CPU.
+
+Since the model is fairly large, you are going to need either **a Colab's high RAM VM** (which requires Pro subscription) or **a Kaggle VM**.
```
Comment (pavithraes) [inline docs/JAX_porting_HF_model.md:24]: I think it'll be helpful to include docs links for both of these setups.

File: docs/JAX_porting_HF_model.md [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+        for layer in self.layers:
+            x = layer(x, position_ids)
+        x = self.norm(x)
+        logits = self.lm_head(x)
+        return logits
+```
+
++++ {"id": "gBfFRlNx1b8Z"}
+
+## Run it
```
Comment (pavithraes) [inline docs/JAX_porting_HF_model.md:336]:
```suggestion
## Run the Flax model
```

File: docs/JAX_porting_HF_model.md [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    print(f"Generated token: {next_token[0]}")
+
+print(tokenizer.decode(input_ids[0]))
+```
+
++++ {"id": "QLAM95pf5RMu"}
+
+There you have it. We have successfully converted the Hugging Face model weights from the safetensors file, load them up in our JAX model and run the model.
+
+For simplicity, we have left out many optimizations (JIT, batch inference, KV cache, leveraging accelerators, SPMD and etc.) to speed things up. Feel free to implement them as an exercise.
```
Comment (pavithraes) [inline docs/JAX_porting_HF_model.md:372]:
```suggestion
There you have it. We have successfully converted the Hugging Face model weights from the safetensors file, loaded them up in our JAX model, and run the model.

For simplicity, we have left out many optimizations (JIT, batch inference, KV cache, leveraging accelerators, SPMD and etc.) to speed things up. Feel free to implement them as an exercise.
```

File: docs/JAX_porting_HF_model.md [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+Note that the weights are stored in bfloat16.
+
++++ {"id": "vQ7Bmvj70jjA"}
+
+## Define the model
+
++++ {"id": "HdHQwfSO1xfY"}
+
+Now we can define the model. This [diagram](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/_images/transformer_vs_llama.svg) from NVidia visualizes the model architecture pretty nicely.
+
```
Comment (pavithraes) [inline docs/JAX_porting_HF_model.md:151]: Not sure if this is a helpful reference, please feel free to include/exclude as you see fit.:)

```suggestion

You can check out the [Flax model implementation in the Transformer's library](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_flax_llama.py) for reference.

```

**Review by pavithraes — COMMENTED** [review]
@windmaple Thanks for opening a PR, I've added some suggestions around narration and running the tutorial. :)

Additionally, since we need an HF access token for execution, we may need to add this tutorial to `nb_execution_excludepatterns` for a successful sphinx build.

**Review by jakevdp — COMMENTED** [review]
I think we'll need to add this to `nb_execution_excludepatterns` in `conf.py`, because execution is failing in the readthedocs environment.

Alternatively, we could add `transformers` and `huggingface_hub` to the test requirements, and that may allow it to run in the readthedocs CI.

File: docs/JAX_porting_HF_model.md [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+    bb6cee0552994adfaefeab4ffafc59bc, cbe3094b322140ee9bbf1bac829ab963, 84998fc5b9014df3b43bc42c729152a0,
+    5fd802b91c28472693ead7723cdedd5f, 265d35cf0dae4d63a7b4f2a7b160098b, d2c13617a5e84498b01ca97370e9a825,
+    cc30646ad6e443318977e939cb43be25, e749c6a737bc4168a0e280335a4639cb, 5f440b28cae6429f8614a2a930867230,
+    ad00a584a1484b1aa6295a4395f4c488, 2b018b81a8a14ff8a103f7395ce65147, 7501a49606b346e997647375b1660377,
+    a6ee1a55467b4027ab526683ce1debe9, e6e24411eac94287aa7bb1ffb6e15d0d, 3edfed06e8f24ccea0ffa6386244c05a,
+    23b0ae2153764e39a737abd5c80d3d21, 6d4aa5dfa939462b9f53a56c41803c0e, 2d81309748f34ad28e0c2f7fc2734c8f,
+    3a75b8e419004b36a06e556fa5a4e535, 48d5fa17a0c3444795708ac8cb5e7f45, b21b452dbdbe4db1a6a8c7ad75553b1b,
+    11e528bfe36341adbc264a6ca05d09d9, ddc376976d7f41b6a5dab76c32f62d57, ed4dc67131144c228f088202197e7015,
+    33ae1220d30d46c19893ed23c230995b, 2dd3319a37fe426daed81f7f3ebba034, ac443b375fe448c3bf028755064e0cc6,
+    98454d975f5842eeb4a4e11173e272fd, 990cb01cfa0c44b4966c45358e516076]
```
[the remark this answers is not in the document]
Comment (windmaple) [reply docs/JAX_porting_HF_model.md:113]: These are not weights. They seem to be a widget aut-generated by HF Hub and are just metadata in a sense.

File: docs/JAX_porting_HF_model.md (hunk 4 above)
Comment (windmaple) [reply docs/JAX_porting_HF_model.md:155]: This is already mentioned in the section 'load the model weights' above.

File: docs/JAX_porting_HF_model.md (hunk 1 above)
Comment (windmaple) [reply docs/JAX_porting_HF_model.md:26]: There isn't much to link to actually, only the Colab pricing page.

**Comment by jakevdp** [comment]
To resolve the build error, add the markdown file to the list here: https://github.com/jax-ml/jax-ai-stack/blob/47125fea7257a00cd3514ccffde673ab873f5947/docs/conf.py#L49

File: docs/conf.py [hunk 6]
```diff
@@ -51,6 +51,7 @@
     'getting_started_with_jax_for_AI.md',
     'JAX_for_PyTorch_users.md',
     'JAX_porting_PyTorch_model.md',
+    'JAX_portting_HF_model.md',
```
Comment (jakevdp) [inline docs/conf.py:54]: Typo: extra `t`

File: docs/conf.py [hunk 7]
```diff
@@ -82,6 +83,7 @@
 nb_execution_excludepatterns = [
     'JAX_for_PyTorch_users.ipynb',
     'JAX_porting_PyTorch_model.ipynb',
+    'JAX_portting_HF_model.ipynb',
```
Comment (jakevdp) [inline docs/conf.py:86]: Typo: extra `t`

File: docs/tutorials.md [hunk 8]
```diff
@@ -10,6 +10,7 @@ The following tutorials are meant as an intro to the full stack:
 getting_started_with_jax_for_AI
 digits_vae
 JAX_for_PyTorch_users
+JAX_portting_HF_model
```
Comment (jakevdp) [inline docs/tutorials.md:13]: Typo: extra `t`

**Comment by jakevdp** [comment]
Hi - thanks for this! We'll need to rebase this against main, becuase the doc files are now in `docs/source`. Thanks!

**Comment by windmaple** [comment]
Oh, no. I think I messed up rebasing. Now it touches all files in `docs/source`.

I'll just send another clean PR.

**Comment by jakevdp** [comment]
You just need to rebase on the updated main branch, and then the diffs will be correct.

**Comment by windmaple** [comment]
It's beyond repair (some of the conflicts are not even from my change).

Easier to start from clean slate.

**Pull request closed without merging**